keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -40,74 +56,145 @@ For operators who want to regularly rotate the signing and verifying keys for pr
+- External signers will only return signature and header while the token will be assembled by TokenGenerator implementation for external signing. This will help ApiServer control the payload and thus the claims.
 
+## Design Details
+
+### KeyService RPC
+
+will be served on a socket as configured via the `--key-service-url` flag.
+
+```proto
+// This service is served by a process on a local Unix Domain Socket.
 service KeyService {
-  // Sign an incoming payload
-  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
-  // List all active public keys
-  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+  // Sign takes a serialized JWT payload, and returns the serialized header and
+  // signature.  The caller can then assemble the JWT from the header, payload,
+  // and signature.
+  //
+  // The plugin MUST set a key id in the returned JWT header.
+  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}
+
+  // FetchKeys returns the set of public keys that are trusted to sign
+  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
+  //
+  // * Every time it tries to validate a JWT carrying an unknown key ID, and
+  //
+  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
+  //   JWKs endpoint.
+  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
 }
 
-message SignPayloadRequest {
-  // payload is the content to be signed. JWT headers must be included by the caller
-  bytes payload = 1;
-  // algorithm specifies which algorithm to sign with
-  string algorithm = 2;
+message SignJWTRequest {
+  // The raw payload to be signed.  A serialized JSON object, no base64 wrapping.
+  optional bytes payload = 1;
 }
-message SignPayloadResponse {
-  // content returns the signed payload
-  bytes content = 1;
+
+message SignJWTResponse {
+  // The header for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  optional bytes header = 1;
+
+  // The signature for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  optional bytes signature = 2;
 }
 
+message FetchKeysRequest {
+    optional string triggering_key_id = 1;
+}
 
-message PublicKey {
-  // public_key is a PEM encoded public key
-  bytes public_key = 1;
-  // certificate is a concatenated list of PEM encoded x509 certificates
-  bytes certificates = 2;
-  // key_id is the key's ID
-  string key_id = 3;
-  // algorithm states the algorithm the key uses
-  string algorithm = 4;
+message FetchKeysResponse {
+  repeated Key keys = 2;
+
+  // The timestamp when this data was pulled from the authoritative source of
+  // truth for verification keys.
+  optional google.protobuf.Timestamp data_timestamp = 3;
 }
```

Comment:
We also wanted to add a `suggest_next_fetch_at` timestamp field, to allow the plugin to control the polling frequency.

Reply:
Added